net-misc/ntpsec: change to single through multi-target build

JamesB192 opened the pull request:
Change to an implementation that only compiles all the code ONCE using the upstream default of a C library with a Python ctypes wrapper. Build the initial install of all the module copies via flit. Replace the upstream egg-info file with the output of flit.

bug: https://bugs.gentoo.org/922819
bug: https://bugs.gentoo.org/910009

<!-- Please put the pull request description above -->

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

JamesB192 commented:
I was going to replace the ebuild with versions using the distutils-r1 eclass, but I still need to figure out how to. However, I did get rid of the egg-info file.

I am not going to change the `UnnecessarySlashStrip` and cannot change the `CERTIFICATE_VERIFY_FAILED` directly. Everything else is preexisting. Although the `UnderscoreInUseFlag` entries are all mine from way back, I was trying to get a deprecated thing going. The waf version shipped with 1.2.2a blocks the `PythonCompatUpdate`, and I'd probably wait until Python 3.13 is stable on amd64 before updating later versions.

```console
f1af0b5c76d1 /gentoo/net-misc/ntpsec # pkgcheck scan --net
net-misc/ntpsec
  UnderscoreInUseFlag: USE flag 'rclock_arbiter' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_generic' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_gpsd' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_hpgps' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_jjy' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_local' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_modem' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_nmea' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_oncore' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_pps' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_shm' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_spectracom' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_trimble' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_truetime' uses reserved underscore character
  UnderscoreInUseFlag: USE flag 'rclock_zyfer' uses reserved underscore character
  UnstableOnly: for arch: [ x86 ], all versions are unstable: [ 1.2.2a-r1, 1.2.3-r1 ]
  RedundantVersion: version 1.2.2a-r1: slot(0) keywords are overshadowed by version: 1.2.3-r1
  PotentialStable: version 1.2.3-r1: slot(0), stabled arches: [ amd64, arm, arm64 ], potential: [ ~x86 ]
  PythonCompatUpdate: version 1.2.2a-r1: PYTHON_COMPAT updates available: python3_12, python3_13
  UnnecessarySlashStrip: version 1.2.2a-r1: ${D%/} unnecessary slash strip on line: 159
  PythonCompatUpdate: version 1.2.3-r1: PYTHON_COMPAT update available: python3_13
  SSLCertificateError: version 1.2.3-r1: SRC_URI: SSL cert error: HTTPSConnectionPool(host='ftp.ntpsec.org', port=443): Max retries exceeded with url: /pub/releases/ntpsec-1.2.3.tar.gz (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1010)'))): https://ftp.ntpsec.org/pub/releases/ntpsec-1.2.3.tar.gz
  SSLCertificateError: version 1.2.3-r1: SRC_URI: SSL cert error: HTTPSConnectionPool(host='ftp.ntpsec.org', port=443): Max retries exceeded with url: /pub/releases/ntpsec-1.2.3.tar.gz.asc (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1010)'))): https://ftp.ntpsec.org/pub/releases/ntpsec-1.2.3.tar.gz.asc
  UnnecessarySlashStrip: version 1.2.3-r1: ${D%/} unnecessary slash strip on line: 158
  PythonCompatUpdate: version 9999: PYTHON_COMPAT update available: python3_13
  UnnecessarySlashStrip: version 9999: ${D%/} unnecessary slash strip on line: 158
```

thesamesam commented on net-misc/ntpsec/ntpsec-9999.ebuild:
```diff
@@ -164,12 +153,18 @@ src_install() {
-python_install() {
+	local impl
+	for impl in "${_PYTHON_SUPPORTED_IMPLS[@]}"; do
+		local instpath=${d#${D%/}}
```
You mentioned you're not going to change it, but I'm not clear on why. In EAPI 7+, it's [guaranteed](https://mgorny.pl/articles/the-ultimate-guide-to-eapi-7.html#d-ed-root-eroot-no-longer-have-a-trailing-slash) to not have a trailing slash.

JamesB192 replied:
It's also unused. I am just too goofy right now to read it. I have commit dropping them and I'll push it tomorrow probably.

thesamesam commented on net-misc/ntpsec/ntpsec-9999.ebuild:
```diff
@@ -164,12 +153,18 @@ src_install() {
+		instpath=/${instpath##/}
+		has "${impl}" "${PYTHON_COMPAT[@]}" && \
+		use "python_targets_${impl}" && \
+		mkdir -pv "${D}/usr/lib/${impl/_/.}/site-packages/ntp/" && \
```
Please use the [helpers](https://devmanual.gentoo.org/eclass-reference/python-utils-r1.eclass/index.html) from python-utils-r1.eclass.

JamesB192 replied:
I think it would be great if I could replace a bunch of my cargo cult glue with something like the following:

```
python_foreach_impl \
    ${EPYTHON} -m gpep517 \
    install-wheel "$(wheel_name}" \
    --optimize all --destdir "${D}"
```

I do not think I can because that would cd into another environment and futz several environment variables.

thesamesam commented on net-misc/ntpsec/ntpsec-9999.ebuild:
```diff
@@ -119,25 +117,16 @@ src_configure() {
-	waf-utils_src_compile check
+src_compile() {
+	waf-utils_src_compile
+	(cd build/main&&${PYTHON} -m build -n)
```
Using `build` seems off to me. We're going _from_ using distutils-r1 (which can use `gpep517`, which is what we use as our preferred frontend) to using a custom call for `build`.

I think you should try doing `DISTUTILS_USE_PEP517=setuptools`?

JamesB192 replied:
Probably not today.

thesamesam commented on net-misc/ntpsec/ntpsec-9999.ebuild:
```diff
@@ -164,12 +153,18 @@ src_install() {
-
-python_install() {
+	local impl
+	for impl in "${_PYTHON_SUPPORTED_IMPLS[@]}"; do
```
This is an internal variable, please don't use it. See other comments for how to replace it.

JamesB192 committed:
net-misc/ntpsec: Replace USE_EXPANDish flags with normal ones

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 commented:
If you don't have an origin/HEAD reference in your git repository, running `pkgcheck --net --commits ${ref}` will result in a misleading error message, which is not ideal.

Also, I would like to understand why the recommended glue scripting is superior for the ebuilds compared to what I currently use.

JamesB192 committed:
net-misc/ntpsec: Replace egg-info with a Flit pyproject.toml file

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 committed:
net-misc/ntpsec: Replace n(+1) used whole tree builds with 1

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 committed:
net-misc/ntpsec: Reduce the number of actual builds to 1 from n(+1)

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 committed:
net-misc/ntpsec: Change to PEP517 build, finally.

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 commented:
```
net-misc/ntpsec
  UnstableOnly: for arch: [ x86 ], all versions are unstable: [ 1.2.2a-r1, 1.2.3-r1 ]
  PythonCompatUpdate: version 1.2.2a-r1: PYTHON_COMPAT updates available: python3_12, python3_13
  RedundantVersion: version 1.2.2a-r1: slot(0) keywords are overshadowed by version: 1.2.3-r1
  RequiredUseDefaults: version 1.2.2a-r1: profile: 'default/linux/amd64/23.0' (298 total) failed REQUIRED_USE: ( python_targets_python3_10 || python_targets_python3_11 )
  PotentialStable: version 1.2.3-r1: slot(0), stabled arches: [ amd64, arm, arm64 ], potential: [ ~x86 ]
  PythonCompatUpdate: version 1.2.3-r1: PYTHON_COMPAT update available: python3_13
  PythonCompatUpdate: version 9999: PYTHON_COMPAT update available: python3_13
```

JamesB192 committed:
net-misc/ntpsec: Use later waf to Python targets 3.11, 3.12

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 committed:
net-misc/ntpsec: Solve conflicting modules problem

gpep517 install-wheel throws a fit if run after
waf-utils_src_install for the same Python.

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

JamesB192 committed:
net-misc/ntpsec: Run tests conditionally

I am requiring FEATURES+=" test" and USE+=" test" for this.

Signed-off-by: James Browning <JamesB.fe80@gmail.com>

thesamesam commented:
Thanks!

I applied some changes on top in:
* 6bca5783e91c687ea600bfaab02879ca9cd3b4d9
* 654bd88912eecc78138bf8ea855f615e701020da
* d86bc0181c2ae3ed0391fab1c66d27a43faaa709
* a7a550ba59e0fe463675c8602f4685ef4aad5040
